WalletWasabi/Blockchain/TransactionBuilding/ChangelessTransactionCoinSelector.cs

```diff
@@ -0,0 +1,46 @@
+using NBitcoin;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using WalletWasabi.Blockchain.TransactionOutputs;
+
+namespace WalletWasabi.Blockchain.TransactionBuilding;
+
+public static class ChangelessTransactionCoinSelector
+{
+	public static bool TryGetCoins(List<SmartCoin> availableCoins, FeeRate feeRate, long target, [NotNullWhen(true)] out List<SmartCoin>? selectedCoins, CancellationToken cancellationToken)
+	{
+		selectedCoins = null;
+		// Keys are effective values of smart coins in satoshis.
+		var sortedCoins = availableCoins.OrderBy(x => x.EffectiveValue(feeRate).Satoshi);
```

Comment:
This is a bug, I think, it should be sorted like this:

```suggestion
		var sortedCoins = availableCoins.OrderByDescending(x => x.EffectiveValue(feeRate).Satoshi);
```

otherwise `foreach` on L28 does not work correctly as that foreach supposes the ordering is decreasing.

(https://github.com/zkSNACKs/WalletWasabi/pull/7028/files#diff-a6aad62bde64d36992014d335ee0a8e8c954d17aca04636270e86525bf50b129R17)

Test in my PR shows what happens when this is wrong - https://github.com/zkSNACKs/WalletWasabi/pull/7028/files#diff-fd3bf14028281cf489dcf30e752c270b013407159078c11bacc5f3aed3407da0R18 - it breaks badly.

Reply:
Anyway, I think we should return results from `TryGetExactMatch` in decreasing order (actually document that) and test that.

Reply:
Absolutely right. I'm not sure why it wasn't in desc order in the beginning.

Reply:
>  we should return results from `TryGetExactMatch` in decreasing order

Hmm, can you elaborate on this? 

> that foreach supposes the ordering is decreasing.

Why?

Reply:
> > we should return results from `TryGetExactMatch` in decreasing order
> 
> Hmm, can you elaborate on this?

It does not matter if it is increasing or decreasing but we should just pick one way or the other so that we can then match long values with the smartcoin values using one foreach.

smart coins: 10, 9, 5, 5, 4, 3, 1
solution: 5, 5, 3, 1

one goes through smart coins and only moves right and when ith smartcoin corresponds with the current solution element, then you know that smartcoin should be in the result. So you have this basic algorithm of having two pointers and only moving them right, so it's O(n).